thinc/optimizers.py

```diff
@@ -354,4 +375,201 @@ def _adam(self, ops, weights, gradient, lr_scale, key, nr_upd):
         )
 
 
-__all__ = ["Adam", "RAdam", "SGD", "Optimizer", "ADAM_DEFAULTS", "SGD_DEFAULTS"]
+class SWA(OptimizerABC):
+    """
+    After the optimization ran 'start_step' number of steps
+    it switches the learning-rate of the optimizer and starts
+    recording the moving averages of all weights.
+    """
+    optimizer: Optimizer
+    start_step: int
+    freq: int
+    learn_rate: FloatOrSeq
+    nr_update: Dict[Tuple[int, str], int]
+    run_swa: bool
+    averages: Dict[Tuple[int, str], FloatsXd]
+
+    def __init__(
+        self,
+        optimizer: Optimizer,
+        learn_rate: FloatOrSeqOrStr,
+        *,
+        freq: int = 1,
+        start_step: int = 0,
+        on_plateau: bool = False
+    ):
+        """
+        optimizer: The inner optimizer whose iterates to average.
+        learn_rate: Learning-rate (or schedule) to run with after
+            the averaging is turned on.
+        freq: Every 'freq' step the moving-averages will be updated.
+        start_step: Start the averaging after start_step number of updates.
+        """
+        self.optimizer = optimizer
+        self.start_step = start_step
+        self.freq = freq
+        self.learn_rate = learn_rate
+        self.nr_update = defaultdict(int)
+        self.run_swa = False
+        self.averages = {}
+        if isinstance(self.learn_rate, str):
+            if self.learn_rate != "constant":
+                raise ValueError(
+                    "The only string option that can be "
+                    "provided as learning-rate for SWA is"
+                    " 'constant' currently."
+                )
+
+    def _update_averages(
+        self,
+        key: Tuple[int, str],
+        weights: FloatsXd,
+    ):
+        """
+        Update moving average of weights.
+        """
+        if key not in self.averages:
+            ops = get_array_ops(weights)
+            self.averages[key] = ops.alloc(weights.shape, dtype="float32")
+            self.averages[key] += weights
+        else:
+            self.averages[key] *= self.nr_update[key]
```

Comment:
I wonder if it makes sense to change this a bit:

* Make a private member `_weight_sums` and only add the weights here.
* Add a property `averages` that returns a dict with `_weight_sums` divided by `nr_update`.

Benefits:

* During training updates are done more often than getting the averages (I think). And you avoid do one multiplication and division less.
* You lose less precision as a result of repeatedly multiplying and dividing (though probably not much of an issue in neural nets).

Downside:

* Takes more memory when retrieving the averages, since you have both the sums and averages in memory.

It's just an idea that we might consider.